File: backend/app/models/business.py
"""Business (tenant) and its settings.

NOTE — no ``from __future__ import annotations`` in this module, deliberately.
That import stringifies EVERY annotation, including ``list["User"]`` on the
Relationship below. SQLModel then cannot see the ``list[...]`` wrapper (it gets one
opaque string), so it hands SQLAlchemy the literal text ``list['User']`` as the
relationship target and mapper configuration dies with:

    InvalidRequestError: expression "relationship("list['User']")" seems to be
    using a generic class as the argument to relationship()

Without the future import, ``list["User"]`` evaluates to ``list[ForwardRef('User')]``
-- SQLModel unwraps the list, and SQLAlchemy resolves the "User" forward reference
against its class registry. Every model module that declares a Relationship must
therefore avoid the future import. Modules with no Relationship (customer, catalog,
file, ...) keep it.
"""

from decimal import Decimal
from typing import TYPE_CHECKING, Any

from sqlalchemy import JSON, Column
from sqlmodel import Field, Relationship

from app.models.base import BaseEntity
from app.models.enums import ReminderAudience, RetentionPolicy

if TYPE_CHECKING:
    from app.models.user import User


class Business(BaseEntity, table=True):
    """A tenant. Every domain row hangs off exactly one of these.

    ARCHITECTURE NOTE: settings are split across three groups (profile, reminders,
    storage/retention) but kept on ONE row rather than in a settings table. A
    business always loads all of its settings together, so a join would buy
    nothing, and a single row keeps the update path atomic.
    """

    __tablename__ = "business"

    # --- Identity -----------------------------------------------------------
    name: str = Field(index=True, max_length=160)
    slug: str = Field(unique=True, index=True, max_length=180)
    description: str | None = Field(default=None, max_length=2000)

    # NOTE: deliberately NOT a foreign key.
    #
    # FileAsset.business_id already points at Business, so a FK here would close a
    # cycle: business -> file_asset -> business. SQLite tolerates that; PostgreSQL
    # cannot create the two tables in any valid order and CREATE TABLE fails
    # outright -- which would break the very Postgres migration path this project
    # promises. (Alembic warns: "unresolvable cycles between tables".)
    #
    # The alternative -- deferrable constraints via use_alter=True -- adds real
    # complexity for a reference that the application already manages: StorageService
    # reference-counts every attach/detach, and the orphan sweep reclaims anything
    # unreferenced. The integrity this FK would enforce is enforced in code, on a
    # path that has to exist anyway.
    logo_file_id: str | None = Field(default=None, index=True, max_length=32)

    # --- Contact ------------------------------------------------------------
    email: str | None = Field(default=None, index=True, max_length=255)
    phone: str | None = Field(default=None, max_length=40)
    whatsapp_number: str | None = Field(default=None, max_length=40)
    website: str | None = Field(default=None, max_length=255)

    # --- Social -------------------------------------------------------------
    facebook_url: str | None = Field(default=None, max_length=255)
    instagram_url: str | None = Field(default=None, max_length=255)
    tiktok_url: str | None = Field(default=None, max_length=255)

    # --- Location -----------------------------------------------------------
    address: str | None = Field(default=None, max_length=500)
    city: str | None = Field(default=None, max_length=120)
    country: str | None = Field(default=None, max_length=120)
    google_maps_url: str | None = Field(default=None, max_length=500)
    latitude: float | None = Field(default=None)
    longitude: float | None = Field(default=None)

    # --- Localisation -------------------------------------------------------
    # Defaults target the primary market (Bhutan). Every one of these is editable per
    # business in Settings -> Localisation, so a shop elsewhere just changes them.
    #
    # currency_symbol is stored separately and is NOT cosmetic: Intl renders BTN as
    # "BTN 1,234.50" under en-US, and only reaches "Nu." under dz-BT, which also
    # switches the digits to Tibetan numerals. The frontend therefore takes the
    # grouping from `locale` and the symbol from this column. See frontend/src/lib/format.ts.
    currency: str = Field(default="BTN", max_length=3)          # ISO-4217
    currency_symbol: str = Field(default="Nu.", max_length=8)
    timezone: str = Field(default="Asia/Thimphu", max_length=64)  # IANA
    locale: str = Field(default="en", max_length=10)
    tax_percentage: Decimal = Field(default=Decimal("0"), max_digits=5, decimal_places=2)

    # Opening hours: {"mon": {"open": "09:00", "close": "18:00", "closed": false}, ...}
    # JSON rather than a WorkingHours table -- it is read as an opaque blob by the
    # UI, never queried by field, so a table would be seven joins for nothing.
    working_hours: dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSON))

    # --- Reminder preferences ----------------------------------------------
    reminders_enabled: bool = Field(default=True)
    # Days *before* the due date to fire a reminder. Spec asks for 1/3/7 + custom;
    # a list makes "custom" free rather than three booleans plus an escape hatch.
    reminder_days_before: list[int] = Field(
        default_factory=lambda: [7, 3, 1], sa_column=Column(JSON)
    )
    reminder_audience: ReminderAudience = Field(default=ReminderAudience.BOTH, max_length=16)
    reminder_send_hour: int = Field(default=9)   # business-local hour, 0-23
    notify_owner_on_overdue: bool = Field(default=True)
    notify_owner_on_payment: bool = Field(default=True)

    # --- Email identity (per-business, overrides the platform default) ------
    email_from_name: str | None = Field(default=None, max_length=120)
    email_reply_to: str | None = Field(default=None, max_length=255)
    email_signature: str | None = Field(default=None, max_length=1000)
    brand_color: str = Field(default="#4F46E5", max_length=9)   # #RRGGBB[AA]

    # --- W3Forms access key (per-business, NOT in the environment) -----------
    #
    # WHY THIS ONE CREDENTIAL LIVES IN THE DATABASE AND SMTP'S DOES NOT
    # -----------------------------------------------------------------
    # It looks inconsistent to keep SMTP_PASSWORD in the environment and this in a
    # table. The difference is what the credential MEANS.
    #
    # An SMTP password is a login to a shared transport; the recipient is carried in
    # the message. One relay can serve every business, so the secret is deployment
    # config and belongs in the environment.
    #
    # A W3Forms access key has no recipient parameter at all -- the key IS the
    # destination inbox. So a single key in the environment would deliver EVERY
    # business's owner notifications to ONE inbox: the deployment owner's. Tenant A's
    # overdue alerts would land in tenant B's mail. On a multi-tenant install that is
    # not a preference, it is a tenancy leak, and no amount of env config fixes it.
    # The key has to be per-business because the thing it identifies is per-business.
    #
    # It is never returned raw over the API (see graphql/mappers.py: the field is
    # exposed masked, and is write-only). Falls back to settings.W3FORMS_ACCESS_KEY
    # when a business has not set its own, which keeps a single-tenant install working
    # with nothing but an env var.
    w3forms_access_key: str | None = Field(default=None, max_length=255)

    # --- Retention / storage ------------------------------------------------
    retention_policy: RetentionPolicy = Field(default=RetentionPolicy.DAYS_30, max_length=16)
    retention_notifications_enabled: bool = Field(default=True)
    storage_quota_mb: int = Field(default=500)   # soft cap; warns, does not block

    # --- Platform -----------------------------------------------------------
    is_active: bool = Field(default=True, index=True)

    users: list["User"] = Relationship(
        back_populates="business",
        sa_relationship_kwargs={"foreign_keys": "User.business_id"},
    )

    @property
    def retention_days(self) -> "int | None":
        return RetentionPolicy(self.retention_policy).days

File: frontend/src/features/settings/api/business.ts
"use client";

/**
 * `business` / `updateBusiness`.
 *
 * The Business row backs FOUR different settings screens (Business, Reminders,
 * Data Retention, and — via currency/locale — every money figure in the app), so
 * there is exactly ONE query key for it. A save on the Reminders page must
 * repaint the currency symbol on the Business page, and a single cache entry is
 * what makes that free.
 */

import { useMutation, useQuery, useQueryClient, type UseQueryResult } from "@tanstack/react-query";

import { gqlRequest } from "@/lib/graphql/client";
import { DEFAULT_CURRENCY, DEFAULT_CURRENCY_SYMBOL, DEFAULT_LOCALE } from "@/lib/format";
import type {
  ID,
  ISODateTime,
  Money,
  ReminderAudience,
  RetentionPolicy,
  WorkingHours,
} from "@/types";

/** Mirrors `BusinessType` in docs/schema.graphql, field for field. */
export interface BusinessSettings {
  id: ID;
  name: string;
  slug: string;
  description: string | null;
  logoUrl: string | null;

  email: string | null;
  phone: string | null;
  whatsappNumber: string | null;
  website: string | null;

  facebookUrl: string | null;
  instagramUrl: string | null;
  tiktokUrl: string | null;

  address: string | null;
  city: string | null;
  country: string | null;
  googleMapsUrl: string | null;
  latitude: number | null;
  longitude: number | null;

  currency: string;
  currencySymbol: string;
  timezone: string;
  locale: string;
  taxPercentage: Money;
  workingHours: WorkingHours;

  remindersEnabled: boolean;
  reminderDaysBefore: number[];
  reminderAudience: ReminderAudience;
  reminderSendHour: number;
  notifyOwnerOnOverdue: boolean;
  notifyOwnerOnPayment: boolean;

  emailFromName: string | null;
  emailReplyTo: string | null;
  emailSignature: string | null;
  brandColor: string;

  /**
   * The W3Forms key is WRITE-ONLY: the API never sends it back, so there is no field
   * here to hold it. These two are all the UI gets, and all it needs — whether a key
   * is installed, and enough of its tail ("••••••••a1b2") to recognise which one.
   */
  hasW3formsAccessKey: boolean;
  w3formsAccessKeyHint: string | null;

  retentionPolicy: RetentionPolicy;
  retentionNotificationsEnabled: boolean;
  storageQuotaMb: number;

  isActive: boolean;
  createdAt: ISODateTime;
}

/** Every field of `BusinessUpdateInput`. All optional — send only what changed. */
export interface BusinessUpdateInput {
  name?: string;
  description?: string | null;
  logoFileId?: ID | null;
  email?: string | null;
  phone?: string | null;
  whatsappNumber?: string | null;
  website?: string | null;
  facebookUrl?: string | null;
  instagramUrl?: string | null;
  tiktokUrl?: string | null;
  address?: string | null;
  city?: string | null;
  country?: string | null;
  googleMapsUrl?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  currency?: string;
  currencySymbol?: string;
  timezone?: string;
  locale?: string;
  taxPercentage?: Money;
  workingHours?: WorkingHours;
  remindersEnabled?: boolean;
  reminderDaysBefore?: number[];
  reminderAudience?: ReminderAudience;
  reminderSendHour?: number;
  notifyOwnerOnOverdue?: boolean;
  notifyOwnerOnPayment?: boolean;
  emailFromName?: string | null;
  emailReplyTo?: string | null;
  emailSignature?: string | null;
  brandColor?: string;
  /**
   * Three states, and they are not interchangeable:
   *   undefined -> leave the stored key alone (the form never holds it, so this is
   *                what an ordinary save sends)
   *   ""        -> remove the stored key and fall back to the server's env var
   *   "abc..."  -> replace the stored key
   */
  w3formsAccessKey?: string;
  retentionPolicy?: RetentionPolicy;
  retentionNotificationsEnabled?: boolean;
}

const BUSINESS_FIELDS = /* GraphQL */ `
  fragment BusinessFields on BusinessType {
    id
    name
    slug
    description
    logoUrl
    email
    phone
    whatsappNumber
    website
    facebookUrl
    instagramUrl
    tiktokUrl
    address
    city
    country
    googleMapsUrl
    latitude
    longitude
    currency
    currencySymbol
    timezone
    locale
    taxPercentage
    workingHours
    remindersEnabled
    reminderDaysBefore
    reminderAudience
    reminderSendHour
    notifyOwnerOnOverdue
    notifyOwnerOnPayment
    emailFromName
    emailReplyTo
    emailSignature
    brandColor
    hasW3formsAccessKey
    w3formsAccessKeyHint
    retentionPolicy
    retentionNotificationsEnabled
    storageQuotaMb
    isActive
    createdAt
  }
`;

const BUSINESS_QUERY = /* GraphQL */ `
  ${BUSINESS_FIELDS}
  query Business {
    business {
      ...BusinessFields
    }
  }
`;

const UPDATE_BUSINESS_MUTATION = /* GraphQL */ `
  ${BUSINESS_FIELDS}
  mutation UpdateBusiness($input: BusinessUpdateInput!) {
    updateBusiness(input: $input) {
      ...BusinessFields
    }
  }
`;

export const businessKeys = {
  all: ["business"] as const,
  detail: () => ["business", "detail"] as const,
};

export function useBusiness(): UseQueryResult<BusinessSettings> {
  return useQuery({
    queryKey: businessKeys.detail(),
    queryFn: async () => {
      const data = await gqlRequest<{ business: BusinessSettings }>(BUSINESS_QUERY);
      return data.business;
    },
    // Currency/locale/timezone are read by half the app — keep them warm.
    staleTime: 5 * 60_000,
  });
}

export function useUpdateBusiness() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (input: BusinessUpdateInput) => {
      const data = await gqlRequest<
        { updateBusiness: BusinessSettings },
        { input: BusinessUpdateInput }
      >(UPDATE_BUSINESS_MUTATION, { input });
      return data.updateBusiness;
    },
    onSuccess: (business) => {
      // Seed the cache from the mutation result rather than refetching: the
      // mutation returns the full, authoritative row.
      queryClient.setQueryData(businessKeys.detail(), business);
      // A retention-policy change moves the goalposts for the next sweep.
      void queryClient.invalidateQueries({ queryKey: ["retention"] });
    },
  });
}

export interface MoneyFormat {
  currency: string;
  locale: string;
  symbol: string;
  timezone: string;
}

/**
 * Currency + locale for every formatCurrency() call in the app, straight from the
 * business row. Falls back to the library defaults while the query is in flight,
 * so a table never renders "undefined" in the gap.
 */
export function useMoneyFormat(): MoneyFormat {
  const { data } = useBusiness();
  return {
    currency: data?.currency ?? DEFAULT_CURRENCY,
    locale: data?.locale ?? DEFAULT_LOCALE,
    symbol: data?.currencySymbol || DEFAULT_CURRENCY_SYMBOL,
    timezone: data?.timezone ?? "UTC",
  };
}
